cmd/boulder-janitor/job.go

```diff
@@ -0,0 +1,181 @@
+// using the batchedDBJob's configured work query, purgeBefore, and batchSize.
+// If there is no error the last primary key ID written to the work channel will
+// be returned, otherwise an error result is returned.
+func (j batchedDBJob) getWork(work chan<- int64, startID int64) (int64, error) {
+	var idBatch []int64
+	_, err := j.db.Select(
+		&idBatch,
+		j.workQuery,
+		map[string]interface{}{
+			"startID": startID,
+			"cutoff":  j.purgeBefore,
+			"limit":   j.batchSize,
+		},
+	)
+	if err != nil && err != sql.ErrNoRows {
+		return 0, err
+	}
+	rowsFound := len(idBatch)
+	workStat.WithLabelValues(j.table).Set(float64(rowsFound))
+	if rowsFound == 0 {
+		return startID, nil
+	}
+	var lastID int64
+	for _, v := range idBatch {
+		work <- v
+		lastID = v
+	}
+	return lastID, nil
+}
+
+// cleanResource uses the configured level of parallelism to run go routines
+// that read ID values from the work channel and delete the corresponding table
+// rows. If the batchedDBJob configures a maxDPS rate then it will be enforced by
+// synchronizing the delete operations on a ticker based on the maxDPS.
+// cleanResource will block until all of the worker go routines complete.
+func (j batchedDBJob) cleanResource(work <-chan int64) {
+	wg := new(sync.WaitGroup)
+	deleted := int64(0)
+
+	var ticker *time.Ticker
+	if j.maxDPS > 0 {
+		ticker = time.NewTicker(
+			time.Duration(float64(time.Second) / float64(j.maxDPS)))
+	}
+
+	for i := 0; i < j.parallelism; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for id := range work {
+				if ticker != nil {
+					<-ticker.C
+				}
+				if err := j.deleteResource(id); err != nil {
+					j.log.AuditErrf(
+						"error deleting ID %d from table %q: %s",
+						id, j.table, err)
+					errStat.WithLabelValues(j.table, "deleteResource").Inc()
+				}
+				_ = atomic.AddInt64(&deleted, 1)
+			}
+		}()
+	}
+
+	wg.Wait()
+	j.log.Infof(
+		"deleted a total of %d rows from table %q",
+		deleted, j.table)
+}
+
+// deleteResource peforms a delete of the given ID from the batchedDBJob's
+// table or returns an error.
+func (j batchedDBJob) deleteResource(id int64) error {
+	// NOTE(@cpu): We throw away the sql.Result here without checking the rows
+	// affected because the query is always specific to the ID auto-increment
+	// primary key. If there are multiple rows with the same primary key MariaDB
+	// has failed us deeply.
+	query := fmt.Sprintf(`DELETE FROM %s WHERE id = ?`, j.table)
+	if _, err := j.db.Exec(query, id); err != nil {
+		return err
+	}
+	j.log.Debugf("deleted ID %d in table %q", id, j.table)
+	deletedStat.WithLabelValues(j.table).Inc()
+	return nil
+}
+
+// RunForever starts a go routine that will run forever getting work with
+// getWork and deleting rows with cleanResource.
+func (j batchedDBJob) RunForever() {
+	var id int64
+	work := make(chan int64)
+
+	go func() {
+		for {
+			lastID, err := j.getWork(work, id)
+			if err != nil {
+				j.log.AuditErr(err.Error())
```

Comment:
This probably shouldn't be logged at the audit level.

Reply:
The existing `expired-authz-purger` logs the same class of error at this level as well:
https://github.com/letsencrypt/boulder/blob/cc2754cc576904190ac6291769950712d5dbdb59/cmd/expired-authz-purger/main.go#L237

Reply:
Fixed in fa4045b